### ISSUE TITLE
Writing Chinese text results in some characters falling back to a different font in web app

### ISSUE THREAD

**Issue by LaurenzV** [typst/typst#5040]
### Description

Example:

```
将要方翼転稿選敗経容友表機。発無生以極会知質著陸質上必者教輸。庭菊意司情門佐略駒給県間相申行。文様理能株原位万一条中意中詳破必資地加。清問明田謝暮合住経件故本性勢長夜最。第向表社旅当英佐処石済度聞連。情全知公真三告用右暮共南。焦情穴毎局兵月過初薄当大異経町。覧薄約最快業武朝員持幌豊生御節。在新大注議中警緊予荷需業治客。
```
Result:
<img width="592" alt="image" src="https://github.com/user-attachments/assets/d79b6fdb-924d-4b9b-8cd0-4804b943fc0c">

Not how there are a few characters that are bolder than the rest. This is kind of a web app issue, but I'm still posting it here so I can label with with "urgent", because I think that should definitely be fixed before the next release, as it's a bad user experience. Maybe by changing the default web app fonts so that a font is chosen that supports both, Japanese and Chinese characters.

From what I can tell, the issue is that it uses MoeSong UN by default, which apparently is missing some characters, which are taken from Noto Sans CJK. Maybe we should just always use Noto Sans CJK?

### Typst version

- [ ] I am using the latest version of Typst

**Comment by LaurenzV** [comment]
I guess this was introduced by https://github.com/typst/webapp-issues/issues/380.

**Comment by peng1999** [comment]
> Maybe we should just always use Noto Sans CJK?

Maybe just move Noto Sans CJK to the higher position of fallback list.

**Comment by laurmaedje** [comment]
> Maybe just move Noto Sans CJK to the higher position of fallback list.

There isn't really one. It goes into last resort fallback here, where all bets are kinda off.

**Comment by laurmaedje** [comment]
I'm not sure whether it's helpful to make it slightly less bad. Relying on font fallback will always be worse than having properly configured fonts.

I think there's things we can do (like a better default font list when lang == zh or similar), and better warnings in the web app for users that do not configure the language. But just changing the fallback order feels like a band-aid fix to me.

**Comment by LaurenzV** [comment]
Fair enough. I guess we can transfer this to `webapp-issues` then?

**Comment by laurmaedje** [comment]
Not sure. The web app has little say in font selection. It just provides a FontBook. So this might require changes in Typst.

**Comment by Andrew15-5** [comment]
I agree that a better fallback font selection can be used when specifying `text.lang`/`text.region`.

**Comment by LaurenzV** [comment]
> But just changing the fallback order feels like a band-aid fix to me.

I think it's worth it to provide a band-aid fix for the web app, just so that the default user experience for CJK users that are using the web app for the first time isn't affected by this.

**Comment by laurmaedje** [comment]
I'm not sure whether it's really better to make it slightly less bad because then it might be just good enough to not configure a proper font, but still bad.

**Comment by laurmaedje** [comment]
I guess it also depends on whether you just have a bit of Chinese text or a document full of it. In the former case, it might be enough.

**Comment by YDX-2147483647** [comment]
[Maxwell pointed out on the Forum](https://forum.typst.app/t/shouldn-t-typst-select-proper-cjkv-glyph-as-any-other-software-do/8722/6?u=y.d.x) that [fontconfig has a mechanism for manually adjusting the priority of fonts (e.g., `~/.config/fontconfig/fonts.conf`).](https://wiki.archlinux.org/title/Localization/Simplified_Chinese#Chinese_characters_displayed_as_variant_(Japanese)_glyphs) One can put _Noto Serif CJK SC_ before _Noto Serif CJK JP_ by leveraging this mechanism (SC comes after JP in alphabetical order).
It is said that the mechanism is supported by operating systems, browsers, terminals, and many softwares.

Maxwell suggests that Typst should respect fontconfig if `text.font` is not set explicitly.

**Comment by sijow** [comment]
Relying on the local fontconfig setup would mean that the output is not reproducible though... If the user moves the document to the webapp or sends it to a collaborator, it would silently compile to a wrong output. I'd rather have smart warnings that show up until the fonts are properly configured.

**Comment by YDX-2147483647** [comment]
I don't have a strong opinion about the fontconfig proposal[^1], but I want to emphasize that there are two types of reproducibility here:

- same font setup + same `main.typ` + **different platforms** (Windows, Linux, typst.app) ⇒ same result
- same font setup + **different tools** (Typst, LaTeX, matplotlib, Firefox, IME) + same platform ⇒ same result

If what Maxwell said is true, then the fontconfig proposal will make it harder to achieve the first type of reproducibility, but make it easier to achieve the second.

I believe that we should not merely focus on the first type of reproducibility. In any case, sharing a Typst document can never be more reproducible than two people in a cloud meeting sharing an online realtime whiteboard, because the latter doesn't even need any font setup.


[^1]: In fact, I'm not familiar with fontconfig. I rarely use Linux with GUI or comprehensive font setup; and on Windows, TeX Live is the only software that I know it uses fontconfig.

**Comment by baptiste** [comment]
Could the web app issue warnings, like it does errors and suggestions? Here it could be "if the user did not set an explicit `set text(lang:)` AND some body text is non-latin AND a font fallback has occurred invisibly" then issue a warning+suggestion that the appearance may be bad and suggest to add `#set text(language: , font: "")` at the top, and maybe also mention fallback?

Just a thought!
